### ISSUE TITLE
Kernel capabilities to run Docker in Hyper

### ISSUE THREAD

**Issue by ayufan** [hyperhq/hyperd#24]
It would be nice to extend kernel config to support: all cgroups, veth, iptables with filter, nat and conntrack.

That would allow to run docker in hyper VM, making the hyper best choice to be used as CI worker.

What do you think?

**Comment by gnawux** [comment]
@ayufan I think in most of the cases, you can directly run docker image with hyper. And I'd like to learn your case, why you want to run docker in hyper.

And actually, you can use your own kernel in the [config file](https://docs.hyper.sh/reference/configuration.html), and reference our [kernel configuration](https://github.com/hyperhq/hyperstart/blob/master/build/kernel_config) . PS: if you want use a customized kernel, you need disable the `Cbfs` config item, or build a new cbfs rom ([reference our make cbfs code](https://github.com/hyperhq/hyperstart/blob/master/build/make-initrd.sh#L35)), and I will add a document page on build your own kernel/initrd.

**Comment by ayufan** [comment]
@gnawux I did it yesterday (recompiled kernel and rebuilt cbfs) and it works great.

About the use case:
Currently Docker-in-Docker is possible, but insecure (requires --privileged). With using full-virtualization docker-in-hyper can work without any security concerns. It's especially useful for CI that for time of build spinups new VM and requires access to Docker in order to build, test and deploy container.

I specifically were thinking about implementing it in https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/ as additional method to run builds.

I guess that your main concern is not inclusion of features, but boot time that can be longer with them enabled. Am I right?

@gnawux One quest: do volumes work in 0.2.1?

**Comment by gnawux** [comment]
@ayufan the volumes work in current version, and if you meet any problems, you can file an issue.

I do not think it will take too much time to boot with a few more options :), just curious about the scenario. If it is a common case for many users, the features should be add to the release package.

**Comment by ayufan** [comment]
Is it common? hard to guess. It would be nice to have, because there's no secure option to run dind :) You could sell it that this is supported in secure manner. There are companies that are fighting with creating selinux, apparmor policies for LXC to have it working and isolated, but still this is not "that" secure as full VM.

**Comment by gnawux** [comment]
True. Thanks for your case @ayufan . 

We will do further investigation, and maybe have all these options in next release.

**Comment by ayufan** [comment]
@gnawux I'll be happy to prepare PR with required kernel features enabled. Just let me know :)

**Comment by Gnep** [comment]
I think it makes sense to support cgroup and other namespaces in pod. So when I create a pod, I have the option to isolate multiple containers in a pod, or share namespace.

The multi-container feature should do the job for @ayufan , though it doesn't require docker daemon in the VM instance.

**Comment by geku** [comment]
Any news on this? I have exactly the same use case and would like to use Hyper pods running a Docker daemon to build Docker images in a secure, isolated way.

**Comment by junoyoon** [comment]
Any progress on this? I have same issues as well.

**Comment by feiskyer** [comment]
@junoyoon Latest docker can be run smoothly inside Hyper pods, just don't forget to mount cgroups before starting dockerd, refer [here](https://github.com/hyperhq/official-images/blob/master/hyperd/cgroupfs-mount).

**Comment by junoyoon** [comment]
@feiskyer  thanks for reply.

I've tried the following steps to test it.

1. run cgroupfs-mount in centos7 PM.
2. install hyperd in centos7 PM.
3. run hyperd container using ```hyperctl run -t centos:latest bash``` 
4. install docker in the hyperd container using 
   * https://docs.docker.com/engine/installation/linux/centos/#/install-with-yum
5. start docker daemon in the hyperd container.

Finally I got the following errors.

```
[root@centos-latest-6240306100 /]# systemctl start docker
Running in chroot, ignoring request.
[root@centos-latest-6240306100 /]#
```

Do I need special steps more for docker daemon?

**Comment by feiskyer** [comment]
@junoyoon Systemd requires much more configurations in containers, try to start dockerd without systemd, e.g. just run `dockerd`

**Comment by junoyoon** [comment]
@feiskyer Thanks... I could run successfully run docker daemon just run ```dockerd```

Then I tried to run docker command ```docker run -it centos:latest```

And I see following result.
```
ERRO[0029] containerd: start container                   error=oci runtime error: container_linux.go:247: starting container process caused "process_linux.go:359: container init caused \"could not create session key: function not implemented\""
                                     id=2dec4e56ab4579541bc35a509cfbbe0b6b70c67c6c0a2a46d629e7ddbb13bf67
                                                                                                        ERRO[0030] Create container failed with error: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:359: container init caused \\\"could not create session key: function not implemented\\\"\"\n"
                                                                                                                                                                  ERRO[0030] Handler for POST /v1.24/containers/2dec4e56ab4579541bc35a509cfbbe0b6b70c67c6c0a2a46d629e7ddbb13bf67/start returned error: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:359: container init caused \\\"could not create session key: function not implemented\\\"\"\n"
docker: Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:359: container init caused \\\"could not create session key: function not implemented\\\"\"\n".
```

Is not  hyperd container compatible docker 1.12?
```
[root@centos-latest-6240306100 /]# docker --version
Docker version 1.12.3, build 6b644ec
```

**Comment by gao-feng** [comment]
@junoyoon Did you update hyper-initrd.img & kernel to the latest version? https://github.com/hyperhq/hyperstart

**Comment by junoyoon** [comment]
@gao-feng  

I've installed the hyperd yesterday following its instruction. https://docs.hypercontainer.io/get_started/install/linux.html
Does the binary distribution outdated?
Do I need to build it from source?

**Comment by gnawux** [comment]
@junoyoon I think you should update hyperhq/hyperstart to the latest master.

Could you confirm this, @gao-feng ?
